spec/controllers/api/v1/packages_controller_spec.rb

```diff
@@ -1596,41 +1611,207 @@ def unpack(qty, pkg, out_of:)
+            expect(package1.reload.on_hand_palletized_quantity).to eq(3)
+            expect(package1.on_hand_boxed_quantity).to eq(5)
+          end
+        end
       end
 
-      it "removes an item from the box" do
-        put :add_remove_item, params: @params1 # add to box
-        put :add_remove_item, params: @params4 # remove it
-        expect(response.status).to eq(201)
-        expect([parsed_body["packages_inventories"]].length).to eq(1)
-        expect(parsed_body["packages_inventories"]["package_id"]).to eq(package1.id)
-        expect(parsed_body["packages_inventories"]["source_id"]).to eq(box.id)
-        expect(parsed_body["packages_inventories"]["source_type"]).to eq("Package")
-        expect(parsed_body["packages_inventories"]["action"]).to eq("unpack")
-        expect(parsed_body["packages_inventories"]["quantity"]).to eq(5)
+      context 'remove item from container' do
+        it "removes an item from the box" do
+          put :add_remove_item, params: @params1 # add to box
+          put :add_remove_item, params: @params4 # remove it
+          expect(response.status).to eq(201)
+          expect([parsed_body["packages_inventories"]].length).to eq(1)
+          expect(parsed_body["packages_inventories"]["package_id"]).to eq(package1.id)
+          expect(parsed_body["packages_inventories"]["source_id"]).to eq(box.id)
+          expect(parsed_body["packages_inventories"]["source_type"]).to eq("Package")
+          expect(parsed_body["packages_inventories"]["action"]).to eq("unpack")
+          expect(parsed_body["packages_inventories"]["quantity"]).to eq(3)
+        end
+
+        context 'when item is removed from box' do
+          it 'updates the on_hand_boxed_quantity' do
+            put :add_remove_item, params: @params1 # add to box
+            put :add_remove_item, params: @params4 # remove it
+            expect(package1.reload.on_hand_boxed_quantity).to eq(2)
+          end
+        end
+
+        context 'when item is removed from pallet' do
+          it 'updates the on_hand_palletized_quantity' do
+            put :add_remove_item, params: @params9 # add to pallet
+            @params9[:task] = 'unpack'
+            @params9[:quantity] = 2
+            put :add_remove_item, params: @params9 # remove it
+            expect(package1.reload.on_hand_palletized_quantity).to eq(1)
+          end
+        end
+
+        context 'when item has both on_hand box and pallet quantity' do
+          before do
+            Package::Operations.pack_or_unpack(container: pallet, package: package1, quantity: 3, location_id: location.id, user_id: user.id, task: 'pack')
+
+            Package::Operations.pack_or_unpack(container: box, package: package1, quantity: 5, location_id: location.id, user_id: user.id, task: 'pack')
+          end
+
+          context 'on removing few items from box' do
+            before do
+              @params1[:task] = 'unpack'
+              @params1[:quantity] = 2
```

Comment:
Kind of the same comment as above, you can use a `let` within the context to define what the variables look like